internal/controller/controller.go

```diff
@@ -311,53 +316,77 @@ func (f *DefaultFanController) UpdateFanSpeed() error {
 	if target > fans.MaxPwmValue {
-		ui.Warning("Tried to set out-of-bounds PWM value %d on fan %s", target, fan.GetId())
+		ui.Warning("Tried to set out-of-bounds PWM value %.2f on fan %s", target, fan.GetId())
 		target = fans.MaxPwmValue
 	} else if target < fans.MinPwmValue {
-		ui.Warning("Tried to set out-of-bounds PWM value %d on fan %s", target, fan.GetId())
+		ui.Warning("Tried to set out-of-bounds PWM value %.2f on fan %s", target, fan.GetId())
 		target = fans.MinPwmValue
 	}
 
 	// map the target value to the possible range of this fan
 	maxPwm := fan.GetMaxPwm()
-	minPwm := fan.GetMinPwm() + f.minPwmOffset
+	minPwm := fan.GetMinPwm()
+	shouldNeverStop := fan.ShouldNeverStop()
+
+	// TODO: in theory even pwmTarget could be a float, because f.setPwm() looks for the closest value
+	//       in the pwm map and uses that
+	var pwmTarget int
+
+	if fan.GetConfig().UseUnscaledCurveValues {
+		pwmTarget = int(math.Round(target))
+		if pwmTarget > 0 && pwmTarget < minPwm {
+			// the fan wouldn't spin with this PWM value anyway, so set 0 instead
+			// (might be better for the hardware and preserve energy)
+			pwmTarget = 0
+		}
+	} else {
+		if target < 1.0 && !shouldNeverStop {
+			// if the fan is allowed to stop and the calculated target is about 0, just set PWM 0
+			pwmTarget = 0
+		} else {
+			// other target values are mapped to minPwm..maxPwm
+			// adjust the target value determined by the control algorithm to the operational needs
+			// of the fan, which includes its supported pwm range (which might be different from [0..255])
+			pwmTarget = minPwm + int(math.Round((target/fans.MaxPwmValue)*float64(maxPwm-minPwm)))
+		}
+	}
 
-	// adjust the target value determined by the control algorithm to the operational needs
-	// of the fan, which includes its supported pwm range (which might be different from [0..255])
-	target = minPwm + int((float64(target)/fans.MaxPwmValue)*(float64(maxPwm)-float64(minPwm)))
+	// if this fan should never stop, make sure its target is always at least minPwm+f.minPwmOffset
+	// (f.minPwmOffset is usually 0, but if the fan doesn't start at MinPwm it gets increased)
+	if shouldNeverStop && pwmTarget < minPwm+f.minPwmOffset {
+		pwmTarget = minPwm + f.minPwmOffset
+	}
 
 	if fan.Supports(fans.FeatureRpmSensor) {
 		// make sure fans never stop by validating the current RPM
 		// and adjusting the target PWM value upwards if necessary
-		shouldNeverStop := fan.ShouldNeverStop()
 		if f.lastTarget != nil {
 			lastTarget := *f.lastTarget
 			// TODO: check this logic
 			lastSetPwm, err := f.getLastTarget()
 			if err != nil {
 				ui.Warning("Error reading last set PWM value of fan %s: %v", fan.GetId(), err)
 			}
-			lastSetTargetEqualsNewTarget := lastTarget == target
-			if shouldNeverStop && lastSetTargetEqualsNewTarget {
+			lastSetTargetEqualsNewTarget := lastTarget == pwmTarget
+			if shouldNeverStop && lastSetTargetEqualsNewTarget { // TODO: also check if fan supports reading RPM?
```

Comment:
This whole block is wrapped in 
```golang
if fan.Supports(fans.FeatureRpmSensor) {
    ...
}
```

So this should already be checked.

Reply:
oh right, missed that